### ISSUE TITLE
mixin alternatives

### ISSUE THREAD

**Issue by mcfiredrill** [embermap/ember-data-storefront#349]
Hi! Thanks for this addon, it's been a great help to my projects.

I was wondering if there was any plan or idea to migrate away from the mixin implementation of the FastBoot adapter, so it could be used with native classes more easily.

https://embermap.github.io/ember-data-storefront/docs/guides/fastboot

**Comment by pjcarly** [comment]
I'm interested in this as well.

Not only for FastBoot, but all mixins in general.

**Comment by ryanto** [comment]
Yup that would be great! This isn't a high priority for me now since the apps I work on don't need it, but if someone is motivated to make this happen I'll absolutely merge the PR.

**Comment by billdami** [comment]
Yeah, this is also very desirable for anyone working in a TypeScript-based ember app, where `Ember.Mixin` is the devil :D

**Comment by pjcarly** [comment]
Looking at the explanation of @chriskrycho over at:
https://discuss.emberjs.com/t/best-way-to-replace-mixins/17395/2

This will be a difficult choice, if we want to keep the same functionality as it is working now. 
Our best option is class decorators, or rewrite the logic in a more service-oriented way, breaking all the implementations.

**Comment by ryanto** [comment]
Thanks PJ!

We recently released 0.18 of storefront, which only supports Ember 3.12+. Does that help with the class decorators?

Also, I'm okay with releasing a breaking change, especially if it gets us away from mixins. What would it look like if it were service oriented?

**Comment by chriskrycho** [comment]
Basically, you would inject a `storefront` service, instead of injecting the `store` with `storefront` methods added to it via mixins. Your `storefront` service would then inject and delegate to the Ember Data service in much the same way that the existing methods do. It would make for a very straightforward swap in most cases, actually.

That is exactly what I would recommend here, after thinking about it all morning. Should be a clean and easy transition and it actually makes for a pretty simple implementation as well!

**Comment by pjcarly** [comment]
Yes, I prefer a storefront service as well. it is a lot cleaner (plus I do not like class decorators for reasons described in the forum post).

If we look at the different Mixins, there are:

**LoadableStore**
Easy, when introducing a new Service "storefront" as described by @chriskrycho above

**LoadableModel**
Would these methods be best moved to that service as well?
Something like:
`this.storefront.modelSideload(model: Model, includesString: string, options: object) : Promise<void>` ?

**FastbootAdapter**
I have no experience with anything regarding Fastboot, could someone elaborate?

**Comment by pjcarly** [comment]
Perhaps it is an idea to undeprecate the functions in the storefront service for starters?
And add functionality for sideloading?
